Added OSRS Wiki Sprite Scraper

oliverpatrick opened the pull request:
Added a button in the main default view that opens a new window with a search input. The input can be used to enter the names of sprites you want to receive from the OSRS Wiki. 

When searched they'll be saved in a new folder called sprites located at src/images/bot/sprites/{sprite}.png

If the bank checkbox is selected the sprites will also be added to the bank folder in the images directory.

Working search and download:
![feature_sprite_scrape](https://user-images.githubusercontent.com/65511074/213760944-6e2961f1-08b0-4efc-a47a-3158068ff899.gif)

Fixed logging:
![feature_sprite_scrape2](https://user-images.githubusercontent.com/65511074/213760960-a0a01e7e-cffa-459c-9eaf-bce6a1d0ab81.gif)

Added bank crop functionality:
![feature_sprite_scrape3](https://user-images.githubusercontent.com/65511074/213760968-a6ccfa3f-344d-4967-9013-d5a5807c4124.gif)

Added bank only checkbox:
![image](https://user-images.githubusercontent.com/65511074/213780886-a7278e1f-d0c2-4e31-91c4-e18aec2a3c2d.png)

Fixes OSR-23

skossan commented:
I like it! Shous probably wrap it in a try/except block in case something goes wrong with the http call due to unforeseen reasons. Thoughts?

kelltom commented:
This is so sick haha

We'll definitely be adding this. One suggestion is to change the log to a text entry field. In the unlikely event that the user decides to scrape a bunch of items at once, it won't append directly to the Tkinter view (they aren't scrollable for some unknown reason). Text entry fields _are_ scrollable.

It could be done similarly to the bot log (in `src/view/info_frame.py`). Basically, when text is to be added to the entry field, you quickly set its editability status to `normal`, append lines to the widget, then quickly set the status back to `disabled` so that it doesn't behave like a notepad, as it is designed to. That way, users can scroll through and see every detail of the scraping process (e.g., on which things it failed to locate).

I don't have time tonight but if you want to take a stab at it, that's the gist. Looks awesome!

-- 

Side note, I do have a decent solution for bank searching planned, which may render bank-only sprites unnecessary. Just thought I'd let you know. Can't hurt leaving it as a feature for now though!

oliverpatrick commented:
> This is so sick haha
> 
> We'll definitely be adding this. One suggestion is to change the log to a text entry field. In the unlikely event that the user decides to scrape a bunch of items at once, right now it looks like it'll append directly to the Tkinter view; they aren't scrollable for some unknown reason, so it may eventually bleed off the window. Text entry fields _are_ scrollable.
> 
> It could be done similarly to the bot log (in `src/view/output_log_frame.py`, it's quite simple). Basically, when text is to be added to the entry field, you quickly set its editability status to `normal`, append lines to the widget, then quickly set the status back to `disabled` so that it doesn't behave like a notepad, as it is designed to. That way, users can scroll through and see every detail of the scraping process (e.g., which things it failed to locate).
> 
> I don't have time tonight but if you want to take a stab at it, that's the gist. Looks awesome!
> 
> --
> 
> ~Side note, I do have a decent solution for bank searching planned, which may render bank-only sprites unnecessary. Just thought I'd let you know. Can't hurt leaving it as a feature for now though!~

I've added the changes you need and rebased. Shoot need some tlc before the merge. Let me know if you need anything else! :)

kelltom commented:
Hey! Just looking at it briefly while I have time. I've updated the "scraper.png" image, as the previous one wasn't rendering on my system properly for some reason. I also pushed some auto-formatting, so your code might look a little different, but should work the exact same.

![chrome_slwYYRP6qa](https://user-images.githubusercontent.com/44652363/214456155-1a95775e-4b3a-4fc6-aa81-50827767e7dc.png) 
![chrome_NrEpwqRcTJ](https://user-images.githubusercontent.com/44652363/214456158-737e0b72-15a2-44ee-9fa9-b64b5fb1e63e.png)

What I would recommend is that a new file is made in the `view` folder -- it can be a class that extends from `customtkinter.CTkFrame`. Basically, it'll allow you to write all the UI code pertaining to the scraper in one single file. Basically, you'd just convert the `btn_scraper_clicked()` function into a new class/file, and then this function could be repurposed to spawn the new view class as a new window (3-4 lines of code). Before the weekend, I'll be pushing some code that does a pretty decent example of this, so we'll have something to model around then.

You could also theoretically put the scraper logic directly into this UI file as well. However, I think having it as a utility is smart -- I can see future use cases where the scraper utility could be used directly by bots automatically. Something like...
```markdown
- bot checks if shark image exists in folder...
- shark image doesn't exist: bot calls scraper utility to fetch one
- scraper utility puts it in the correct folder for next time
```
This could be a super huge feature.

One other thing is that the UI seems to stop responding whenever the submit button is clicked. I am curious how that wasn't happening for you in your videos. But don't worry about it, it should be an easy fix with a little bit of threading -- I'll have time this weekend to play with it unless you want to take a stab at that. 

https://user-images.githubusercontent.com/44652363/214454871-63ea67a4-0615-421f-b0e9-20dc174c68b9.mp4

One last thing is that we can prevent the user from erroneously selecting both checkboxes by using radio buttons instead of checkboxes, which only allow a single selection:

![image](https://user-images.githubusercontent.com/44652363/214455060-0b7a1b95-311b-4cbf-88b4-36ff2516e145.png)

I don't think I've used radio buttons anywhere in this project, but luckily Tom @ Custom Tkinter has a [pretty good wiki](https://github.com/TomSchimansky/CustomTkinter/wiki/CTkRadioButton).

----------

I know that was a lot, but there's no rush, and I don't mind picking things up wherever you leave them! I should have time over the weekend to dive in and make this thing official. Until then, feel free to experiment or do whatever. Pumped about this one!

kelltom commented:
Oh the issue I'm having seems to be an isolated problem. The debugger can't even detect the point of failure.

kelltom commented:
This is incompatible with IPv6, so that's something that should go in the release notes. Just putting this here for myself.

kelltom commented:
All that's left is:

- UI improvements
- Modify download function to use a temp path to avoid having to delete normal files during bank_only option

kelltom commented:
The UI has been fixed.

One more thing that needs to be done is the cropping needs to be improved. I just realized that cropping 10px off the top is not a great solution, as the image sits about centred in its bank slot. So, we need to know the exact bank slot dimensions to properly crop the perfect amount off each image.

Almost there!

kelltom commented:
https://user-images.githubusercontent.com/44652363/218287103-8ad30665-e2c9-4100-badc-879158ec34a5.mp4
